Add Timer's Time Remaining to Digitial Watchface

JustScott opened the pull request:
Adds a live output of the timer's time remaining, along with an hourGlass symbol to the left. Both of these are placed above the current time and are the same color as the date as to not draw attention away from the current time. The icon and the time remaining are both set to hidden if the timer isn't running.

![timer_set_to_show](https://github.com/InfiniTimeOrg/InfiniTime/assets/68134791/59518765-c6f4-4960-93cb-8aec093c4024)
![timer_set_to_show2](https://github.com/InfiniTimeOrg/InfiniTime/assets/68134791/2179eabd-c258-4cb8-96ff-f8e3dfb45074)
![timer_set_to_hidden](https://github.com/InfiniTimeOrg/InfiniTime/assets/68134791/e6d2d71e-22be-4808-85ed-22b4952dadab)

Looking for opinions on possible changes to the timer's and symbol's font color and/or position.

vkareh commented:
@JustScott 
> Looking for opinions on possible changes to the timer's and symbol's font color and/or position.

I like this feature a lot, as I use the timer regularly. Since you asked for opinions, how about making it part of the `StatusIcons`? This way you can see it in several other screens. Although maybe this makes it look too busy. Not sure, but just a suggestion.

JustScott commented:
Here's how the timer would look as a StatusIcon:
![timer_in_watchface](https://github.com/InfiniTimeOrg/InfiniTime/assets/68134791/490ad3ed-8d31-4a3c-b8b5-e6d4aab13e7c)
![timer_in_app_selection](https://github.com/InfiniTimeOrg/InfiniTime/assets/68134791/2f249fd3-4a51-4022-a785-19fb1c52119f)


There are currently 2 issues with this:
1. The status icons are only set to update when switching between screens in the `Tile.cpp` file, which means the time remaining will only update once each time you swipe between screens. This could be changed to update more often like is done on the clock faces, but I'm unsure of how this would effect power consumption.
2. The timer is shown on the digital watch face screen both in the status icons and above the current time. We could have just avoided placing it in the watch face itself, and only placed it in the status icons with the change suggested in issue 1 above... but this would now take the same position as the new weather widget. So the obvious solution is just to detect if the current open app is the digital watch face, and set the timer to hidden.

I'll implement these fixes/changes, but first I'd like some feedback on whether this is something people actually want. I personally think it could be nice, but honestly I don't see myself needing to know how much time is remaining on a timer in the second it takes me to switch between apps (but some people may).

vkareh commented:
> The timer is shown on the digital watch face screen both in the status icons and above the current time. We could have just avoided placing it in the watch face itself, and only placed it in the status icons with the change suggested in issue 1 above... but this would now take the same position as the new weather widget. So the obvious solution is just to detect if the current open app is the digital watch face, and set the timer to hidden.

I've been using a version of the digital face with the weather within the face itself:
![InfiniSim_2024-02-01_131052](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/d5de35a3-d4db-48f1-8244-d1900d83b5e3)

I think this change would go nicely with the timer in the status bar.

> The status icons are only set to update when switching between screens in the Tile.cpp file, which means the time remaining will only update once each time you swipe between screens. This could be changed to update more often like is done on the clock faces, but I'm unsure of how this would effect power consumption.

This is a problem. It could be solved by adding decreasing the task refresh period like this (not tested):
```diff
diff --git a/src/displayapp/screens/Tile.cpp b/src/displayapp/screens/Tile.cpp
index 7c585a4b..ffdb9f44 100644
--- a/src/displayapp/screens/Tile.cpp
+++ b/src/displayapp/screens/Tile.cpp
@@ -87,7 +87,7 @@ Tile::Tile(uint8_t screenID,
   btnm1->user_data = this;
   lv_obj_set_event_cb(btnm1, event_handler);
 
-  taskUpdate = lv_task_create(lv_update_task, 5000, LV_TASK_PRIO_MID, this);
+  taskUpdate = lv_task_create(lv_update_task, 500, LV_TASK_PRIO_MID, this);
 
   UpdateScreen();
 }
```
500ms lets you see the timer seconds real time and it shouldn't really affect power consumption _that_ much, I think, since the status icons only show up in screens that support auto-sleep of the display.

JustScott commented:
@vkareh Ah I missed the part where the status icons update every 5 seconds. I think moving this timer to the status icons would be the best solution if the weather icon is moved as shown in your image above. I'll try this out.

JustScott commented:
Alright I moved the timer from right above the time on the digital watchface, to the StatusIcons 'area'. 

Also, I'm VERY new to C++, and was just following what looked to be the right way to pass this timer controller around, but I feel it got added to too many places... so feedback on whether the code is correct or not would be much appreciated. To me it seems like it would be best if I could just define the timer controller within `StatusIcons.cpp` instead of having to pass it to every StatusIcons class call.

vkareh commented:
I created a PR to move the weather widget: https://github.com/InfiniTimeOrg/InfiniTime/pull/1998

JustScott commented:
I've applied the changes from all of vkareh's reviews, and have been using this without issue for quite a while now. Here's some images of how it looks with the digital watch face weather changes. @JF002 would this be something you'd be interested in including in version 1.16.0?

![InfiniSim_2024-12-20_225411](https://github.com/user-attachments/assets/2dd3e82f-79d3-4e86-9dc5-749e453f115b)
![InfiniSim_2024-12-20_225412](https://github.com/user-attachments/assets/d7476ad9-53f8-4baa-885e-1f2a040d8d17)

JustScott commented:
Adjusted the code to work with the new alarm icon in the status icons from #1884. To prevent crowding I moved the timer over a hair:

![InfiniSim_2025-01-18_134142](https://github.com/user-attachments/assets/04691816-9060-4034-8947-6076bb97e5a7)
![InfiniSim_2025-01-18_134501](https://github.com/user-attachments/assets/6c3aaa1c-146c-40d8-8641-b877545f32a7)

JustScott commented:
I've moved the `timeRemaining` and `timerIcon` to there own `timerContainer`, and moved that container into the primary `container` to allow them to dynamically align with the other icons, while being able to place them closer together than the others. When more than two widgets (exluding batteryIcon) are made visible, the `timerContainer` is set to hidden and a `soloTimerIcon` is made visible. Replacing the live timer with a simple icon allows the user to avoid any overlapping with the time in the top left when many icons are visible, while still letting them know a timer is active... and when a few icons go back to being hidden, the live timer will be made visible again.

![12h_with_solo_timer_icon](https://github.com/user-attachments/assets/dfbbaaec-490c-4343-bb83-14d72f5f69a0)
![12h_with_time_remaining](https://github.com/user-attachments/assets/1d0d29f8-0873-492f-8927-403f96c378d9)

I feel the space between the time and the AM and PM is a bit unecessary and ugly, this is how it would look without the space. I can change it in this PR, or it can be left to another PR... just thought I would show the difference.
![12h_no_space_with_time_remaining](https://github.com/user-attachments/assets/06d08a9c-fa8f-487f-955b-a6ff400a6299)

![digital_watchface_solo_timer_icon](https://github.com/user-attachments/assets/91ef021a-5aa3-4232-b243-6b2749d0d507)
![digital_watchface_time_remaining](https://github.com/user-attachments/assets/a9c1dd30-0c0f-499f-be6a-067205dc5a7c)

There could be another icon allowed with the 24 hour time format, but this would require bringing the settingsController in... I'm not opposed to it, but it seems unecessary
![24H_icons_with_time_remaining](https://github.com/user-attachments/assets/f37095eb-9f57-4bdb-9f4f-939a3d3adaa4)
![24H_no_icons_just_timer](https://github.com/user-attachments/assets/5bb936bc-e222-4346-b8b9-938843aed68c)

mark9064 commented:
Maybe to save space the timer could shorten to `XXm`/`XXs`?

JustScott commented:
> Maybe to save space the timer could shorten to `XXm`/`XXs`?

Incredible timing, I was just working on something like this.

JustScott commented:
Minutes are now only shown if greate than 0. Another icon is also allowed if the user is using 24H time instead of 12H time.

![mintues_12h](https://github.com/user-attachments/assets/080f1da3-a75d-4c95-8427-184ab8cd28cc)
![seconds_12h](https://github.com/user-attachments/assets/0b94b286-4f73-441d-b51a-d025210f9424)
![minutes_24h](https://github.com/user-attachments/assets/1c2ec440-2f6c-490c-b576-d9eafcce3924)
![seconds_24h](https://github.com/user-attachments/assets/18be2b49-73da-4618-87dd-0a6c8587eebe)

vkareh commented:
If you want to be more aggressive, I'd even remove the colon `:` when it's just seconds remaining. Seems unnecessary and shaves off a whole character.

This looks great, BTW, and I use this feature every day. Looking forward to rebasing the new look into my daily-driver branch.

JustScott commented:
> If you want to be more aggressive, I'd even remove the colon `:` when it's just seconds remaining. Seems unnecessary and shaves off a whole character.

I'd be okay with this, but want to see what others think as well:
![no_colon](https://github.com/user-attachments/assets/fbcd16a0-21c3-43bb-865a-f7cfe0d370a8)

JustScott commented:
Reduced timer letter spacing to save space

Allows up to four icons (using an example stopWatch icons for the fourth icon):
![normal_letter_spacing](https://github.com/user-attachments/assets/9a280ea9-45ac-444a-b7e0-0de1445c6558)
![with_reduced_letter_spacing](https://github.com/user-attachments/assets/d578cb9c-af80-4e13-91fc-43ef846ce6b6)

Another example with hour support from PR #2243:
![hour_without_reduced_letter_spacing](https://github.com/user-attachments/assets/4bfc6f51-eba1-446d-af86-2667e9fbd8fb)
![hour_with_reduced_letter_spacing](https://github.com/user-attachments/assets/baa57834-350b-4ac6-b519-2a9cd60f7ec5)

minacode commented:
Most aggressive would be to replace the current time with the time remaining for as long as the timer is running 😄
